app/controllers/company/CompanyAddressController.scala

```diff
@@ -56,27 +69,74 @@ class CompanyAddressController @Inject()(override val messagesApi: MessagesApi,
   def onPageLoad(mode: Mode): Action[AnyContent] =
     (identify andThen getData andThen requireData).async {
       implicit request =>
-        getFormToJson(mode).retrieve.right.map(get)
+        getFormToJsonForLoad(mode).retrieve.right.map(get)
     }
 
   def onSubmit(mode: Mode): Action[AnyContent] =
     (identify andThen getData andThen requireData).async {
       implicit request =>
-        getFormToJson(mode).retrieve.right.map(post(mode, _, CompanyAddressPage))
+        getFormToJsonForSubmit(mode).retrieve.right.map(post(mode, _, CompanyAddressPage))
     }
 
-  def getFormToJson(mode: Mode)(implicit request: DataRequest[AnyContent]): Retrieval[Form[Address] => JsObject] =
+  override def post(mode: Mode, json: Form[Address] => JsObject, addressPage: QuestionPage[Address])
+    (implicit request: DataRequest[AnyContent], ec: ExecutionContext, hc: HeaderCarrier, messages: Messages): Future[Result] = {
+
+    form(messages).bind(retrieveFieldsFromRequestAndAddCountryForUK).fold(
+      formWithErrors =>
+        renderer.render(viewTemplate, json(formWithErrors)).map(BadRequest(_)),
+      value =>
+        for {
+          updatedAnswers <- Future.fromTry(request.userAnswers.set(addressPage, value))
+          _ <- userAnswersCacheConnector.save(updatedAnswers.data)
+        } yield {
+          Redirect(navigator.nextPage(addressPage, mode, updatedAnswers))
+        }
+    )
+  }
+
+  private def getFormToJsonForLoad(mode: Mode)(implicit request: DataRequest[AnyContent]): Retrieval[Form[Address] => JsObject] =
```

Comment:
Is there any need to pull the body of the page load and submit methods out into other methods? I assume these aren't reused anywhere else.

Reply:
No I don't think so, it was just like that originally and I left it as was but I can certainly move the code - I will have a look